projects/lights_control/src/lights_gpio.c

```diff
@@ -2,211 +2,113 @@
-      [LIGHTS_SIDE_RIGHT_INDICATOR_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 12 },  //
-      [LIGHTS_RIGHT_TURN_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 10 },            //
-    };
-
-static const GPIOAddress
-    s_rear_signal_left_peripheral_addresses[NUM_LIGHTS_REAR_SIGNAL_LEFT_PERIPHERALS] = {
-      [LIGHTS_LEFT_REAR_OUTER_TURN_PERIPHERAL] = { .port = GPIO_PORT_A, .pin = 10 },  //
-      [LIGHTS_LEFT_REAR_TURN_PERIPHERAL] = { .port = GPIO_PORT_A, .pin = 9 },         //
-    };
-
-static const GPIOAddress
-    s_rear_signal_right_peripheral_addresses[NUM_LIGHTS_REAR_SIGNAL_RIGHT_PERIPHERALS] = {
-      [LIGHTS_RIGHT_REAR_OUTER_TURN_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 0 },  //
-      [LIGHTS_RIGHT_REAR_TURN_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 10 },       //
-    };
-
-static const GPIOAddress s_board_type_address = { .port = GPIO_PORT_B, .pin = 13 };
-
-static LightsBoard s_board;
-
-static const GPIOAddress *s_address_front_lookup[] = {
-  [LIGHTS_EVENT_HORN] = s_horn_peripheral_addresses,                        //
-  [LIGHTS_EVENT_HIGH_BEAMS] = s_high_beams_peripheral_addresses,            //
-  [LIGHTS_EVENT_LOW_BEAMS] = s_low_beams_peripheral_addresses,              //
-  [LIGHTS_EVENT_DRL] = s_drl_peripheral_addresses,                          //
-  [LIGHTS_EVENT_SIGNAL_LEFT] = s_front_signal_left_peripheral_addresses,    //
-  [LIGHTS_EVENT_SIGNAL_RIGHT] = s_front_signal_right_peripheral_addresses,  //
-};
-
-static const GPIOAddress *s_address_rear_lookup[] = {
-  [LIGHTS_EVENT_BRAKES] = s_brakes_peripheral_addresses,
-  [LIGHTS_EVENT_STROBE] = s_strobe_peripheral_addresses,
-  [LIGHTS_EVENT_SIGNAL_LEFT] = s_rear_signal_left_peripheral_addresses,
-  [LIGHTS_EVENT_SIGNAL_RIGHT] = s_rear_signal_right_peripheral_addresses,
-};
-
-static StatusCode prv_init_peripheral(GPIOAddress *peripherals) {
+StatusCode prv_set_peripherals(uint16_t bitset, GPIOState state) {
   uint8_t i;
-
-  // same settings are used for all ports
-  GPIOSettings settings = { .direction = GPIO_DIR_OUT,
-                            .state = GPIO_STATE_HIGH,
-                            .resistor = GPIO_RES_NONE,
-                            .alt_function = GPIO_ALTFN_NONE };
-
-  for (i = 0; i < SIZEOF_ARRAY(peripherals); i++) {
-    StatusCode status = gpio_init_pin(&peripherals[i], &settings);
-    if (status != STATUS_CODE_OK) {
-      return status;
+  while (bitset) {
+    i = __builtin_ffs(bitset) - 1;  // index of first 1 bit
+    StatusCode s = gpio_set_state(&s_addresses[i], state);
+    if (s) {
+      return s;
     }
+    bitset &= ~(1 << i);  // bit is read, so we clear it
   }
   return STATUS_CODE_OK;
 }
 
-StatusCode lights_gpio_init() {
-  uint8_t i;
+static StatusCode prv_get_board_type(LightsBoard *board, LightsConfig *config) {
+  // initializing board type pin
+  StatusCode board_type_init_status =
+      gpio_init_pin(config->board_type_address, config->gpio_settings_in);
+  if (board_type_init_status) {
```

Comment:
Use `status_ok` or the `status_ok_or_return` macro

Reply:
cool